gotham/src/router/builder/single.rs

```diff
@@ -16,6 +16,24 @@ use crate::state::State;
 use core::future::Future;
 use futures::FutureExt;
 
+pub trait AsyncHandlerFn<'a> {
+    type Res: IntoResponse + 'static;
+    type Fut: std::future::Future<Output = Result<Self::Res, HandlerError>> + Send + 'a;
+    fn call(self, arg: &'a mut State) -> Self::Fut;
+}
+
+impl<'a, D, R, F> AsyncHandlerFn<'a> for F
+where
+    F: FnOnce(&'a mut State) -> D,
+    R: IntoResponse + 'static,
+    D: std::future::Future<Output = Result<R, HandlerError>> + Send + 'a,
```

Comment:
this was only called D because the original gist was trying to return `impl Debug + 'a`. Probably want some better names for this.